README.md
[![CI](https://github.com/jbroutier/image-generator/actions/workflows/ci.yml/badge.svg)](https://github.com/jbroutier/image-generator/actions/workflows/ci.yml)
[![PHP version](https://img.shields.io/badge/php-8.0+-787cb5?logo=php)](https://github.com/jbroutier/image-generator)
[![License](https://img.shields.io/github/license/jbroutier/image-generator)](https://github.com/jbroutier/image-generator/blob/main/LICENSE)
[![Packagist](https://img.shields.io/packagist/v/jbroutier/image-generator)](https://packagist.org/packages/jbroutier/image-generator)

# Image generator

Offline random placeholder image generator.

This is a rewrite of [standaniels/image-generator](https://github.com/standaniels/image-generator) which keeps the overall
package goal but adds many features, including:

- More output formats: AVIF (PHP 8.1+), BMP, GIF, JPEG, PNG, WBMP, WebP, XMB.
- More color options.
- More shape options.
- Transparent background support.
- No dependency on `ext-exif`.
- Full PHP 8.0+ compatibility.

## Basic example

```php
<?php

use ImageGenerator\Canvas;
use ImageGenerator\Color;
use ImageGenerator\Shape\Shape;

$canvas = Canvas::create(400, 400);

for ($i = 0; $i < 100; $i++) {
    $canvas->draw(Shape::random(), Color::random());
}

$canvas->render('output.png');
```

The generated image will look like this:

![(Example)](example.png)

## Advanced examples

### Background color

It is possible to define the background color, to use a semi-transparent background or a totally transparent background.

```php
// Fully-transparent background
$canvas->enableTransparency();

// Semi-transparent light background.
$canvas->enableTransparency();
$canvas->fill(Color::create(236, 240, 241, 0.55));

// Fully-opaque dark background
$canvas->fill(Color::create(44, 62, 80, 1));
```

### Shapes

```php
// Circles only
$canvas->draw(Circle::random(), Color::random());

// Hexagons only
$canvas->draw(Polygon::random(['min_sides' => 6, 'max_sides' => 6]), Color::random());

// Stars with 5 to 7 points
$canvas->draw(Star::random(['min_points' => 5, 'max_points' => 7]), Color::random());
```

### Colors

```php
// Semi-transparent shade of blue
$canvas->draw(Shape::random(), Color::create(41, 128, 185, 0.65));

// Fully-opaque shade of green
$canvas->draw(Shape::random(), Color::create(39, 174, 96, 1));
```

### Shape size ratio

The size of shapes is automatically calculated based on the size of the canvas. However, it's possible to adjust the
size ratio to obtain smaller or larger shapes.

```php
// With random shapes
$shape = Shape::random([
    'circle' => [
        'min_ratio' => 0.0625,
        'max_ratio' => 0.125,
    ],
]);

// With a single shape
$shape = Shape::circle([
    'min_ratio' => 0.0625,
    'max_ratio' => 0.125,
]);
```

### Alpha range

It is possible to adjust the range of transparency or turn off transparency completely. A value of 0 corresponds to a
fully transparent color while a value of 1 corresponds to a fully opaque color.

```php
// Full alpha range
$color = Color::random([
    'min_alpha' => 0,
    'max_alpha' => 1
]);

// Fully-opaque colors
$color = Color::random([
    'min_alpha' => 1,
    'max_alpha' => 1
]);
```

### Output format

The default format is PNG but it is possible to choose another image format. Note that AVIF format support is only
available from PHP 8.1. Support for different image formats depends on how PHP was built and may vary depending on your
platform.

```php
$canvas->render('output.avif' , 'avif');
$canvas->render('output.bmp'  , 'bmp' );
$canvas->render('output.gif'  , 'webp');
$canvas->render('output.jpeg' , 'jpeg');
$canvas->render('output.png'  , 'png' );
$canvas->render('output.wbmp' , 'wbmp');
$canvas->render('output.webp' , 'webp');
$canvas->render('output.xbm'  , 'xbm' );
```

### Image quality

Some image formats allow you to adjust the image quality. Options are passed directly to the underlying render function.

```php
$canvas->render('output.avif', 'avif', 80, 5); // Quality: 80, Speed: 5
$canvas->render('output.bmp' , 'bmp' , true ); // Enable compression
$canvas->render('output.jpeg', 'jpeg', 90   ); // Quality: 90
$canvas->render('output.png' , 'png' , 9    ); // Compression level: 9
$canvas->render('output.webp', 'webp', 80   ); // Quality: 80
```


CHANGELOG.md
**Version 1.0.0 (2023-06-01)**

- Initial release.


src/Color.php
<?php

namespace ImageGenerator;

use InvalidArgumentException;

final class Color
{
    private int $red;

    private int $green;

    private int $blue;

    private float $alpha;

    /**
     * Creates a new color.
     *
     * @param int $red The color red component value. Must be between 0 and 255.
     * @param int $green The color green component value. Must be between 0 and 255.
     * @param int $blue The color blue component value. Must be between 0 and 255.
     * @param float $alpha The color alpha value. Must be between 0.0 (fully transparent) and 1.0 (fully opaque).
     */
    public function __construct(int $red, int $green, int $blue, float $alpha)
    {
        $this->setRed($red);
        $this->setGreen($green);
        $this->setBlue($blue);
        $this->setAlpha($alpha);
    }

    /**
     * Returns the color red component value.
     *
     * @return int
     */
    public function getRed(): int
    {
        return $this->red;
    }

    /**
     * Sets the color red component value.
     *
     * @param int $red The color red component value. Must be between 0 and 255.
     *
     * @return Color
     */
    public function setRed(int $red): Color
    {
        if ($red < 0 || $red > 255) {
            throw new InvalidArgumentException('The red component value must be between 0 and 255.');
        }

        $this->red = $red;
        return $this;
    }

    /**
     * Returns the color green component value.
     *
     * @return int
     */
    public function getGreen(): int
    {
        return $this->green;
    }

    /**
     * Sets the color green component value.
     *
     * @param int $green The color green component value. Must be between 0 and 255.
     *
     * @return Color
     */
    public function setGreen(int $green): Color
    {
        if ($green < 0 || $green > 255) {
            throw new InvalidArgumentException('The green component value must be between 0 and 255.');
        }

        $this->green = $green;
        return $this;
    }

    /**
     * Returns the color blue component value.
     *
     * @return int
     */
    public function getBlue(): int
    {
        return $this->blue;
    }

    /**
     * Sets the color blue component value.
     *
     * @param int $blue The color blue component value. Must be between 0 and 255.
     *
     * @return Color
     */
    public function setBlue(int $blue): Color
    {
        if ($blue < 0 || $blue > 255) {
            throw new InvalidArgumentException('The blue component value must be between 0 and 255.');
        }

        $this->blue = $blue;
        return $this;
    }

    /**
     * Returns the color alpha value.
     *
     * @return float
     */
    public function getAlpha(): float
    {
        return $this->alpha;
    }

    /**
     * Sets the color alpha value.
     *
     * @param float $alpha The color alpha value. Must be between 0.0 (fully transparent) and 1.0 (fully opaque).
     */
    public function setAlpha(float $alpha): Color
    {
        if ($alpha < 0 || $alpha > 1) {
            throw new InvalidArgumentException('The alpha value must be between 0 and 1.');
        }

        $this->alpha = $alpha;
        return $this;
    }

    /**
     * Creates a new color.
     *
     * @param int $red The color red component value. Must be between 0 and 255.
     * @param int $green The color green component value. Must be between 0 and 255.
     * @param int $blue The color blue component value. Must be between 0 and 255.
     * @param float $alpha The color alpha value. Must be between 0.0 (fully transparent) and 1.0 (fully opaque).
     *
     * @return Color
     */
    public static function create(int $red, int $green, int $blue, float $alpha): Color
    {
        return new Color($red, $green, $blue, $alpha);
    }

    /**
     * Creates a random color.
     *
     * @param array{
     *     min_red?: int,
     *     max_red?: int,
     *     min_green?: int,
     *     max_green?: int,
     *     min_blue?: int,
     *     max_blue?: int,
     *     min_alpha?: float,
     *     max_alpha?: float
     * } $options An optional array of options. The following options are supported:
     *
     * min_red - The color red component minimum value. Must be between 0 and 255. Defaults to 0.
     * max_red - The color red component maximum value. Must be between 0 and 255. Defaults to 255.
     * min_green - The color green component minimum value. Must be between 0 and 255. Defaults to 0.
     * max_green - The color green component maximum value. Must be between 0 and 255. Defaults to 255.
     * min_blue - The color blue component minimum value. Must be between 0 and 255. Defaults to 0.
     * max_blue - The color blue component maximum value. Must be between 0 and 255. Defaults to 255.
     * min_alpha - The color minimum alpha value. Must be between 0.0 and 1.0. Defaults to 0.6.
     * max_alpha - The color maximum alpha value. Must be between 0.0 and 1.0. Defaults to 0.8.
     *
     * @return Color
     */
    public static function random(array $options = []): Color
    {
        $minRed = $options['min_red'] ?? 0;
        $maxRed = $options['max_red'] ?? 255;
        $red = rand($minRed, $maxRed);

        $minGreen = $options['min_green'] ?? 0;
        $maxGreen = $options['max_green'] ?? 255;
        $green = rand($minGreen, $maxGreen);

        $minBlue = $options['min_blue'] ?? 0;
        $maxBlue = $options['max_blue'] ?? 255;
        $blue = rand($minBlue, $maxBlue);

        $minAlpha = (int)(($options['min_alpha'] ?? 0.6) * 100);
        $maxAlpha = (int)(($options['max_alpha'] ?? 0.8) * 100);
        $alpha = rand($minAlpha, $maxAlpha) / 100;

        return new Color($red, $green, $blue, $alpha);
    }
}


src/Image.php
<?php

namespace ImageGenerator;

final class Image extends \SplFileInfo
{
    private int $width;

    private int $height;

    private string $mimeType;

    /**
     * Creates a new image.
     *
     * @param string $path The file path.
     * @param int $width The image width.
     * @param int $height The image height.
     * @param string $mimeType The image MIME type.
     */
    public function __construct(string $path, int $width, int $height, string $mimeType)
    {
        parent::__construct($path);

        $this->width = $width;
        $this->height = $height;
        $this->mimeType = $mimeType;
    }

    /**
     * Returns the image width.
     *
     * @return int
     */
    public function getWidth(): int
    {
        return $this->width;
    }

    /**
     * Returns the image height.
     *
     * @return int
     */
    public function getHeight(): int
    {
        return $this->height;
    }

    /**
     * Returns the image MIME type.
     *
     * @return string
     */
    public function getMimeType(): string
    {
        return $this->mimeType;
    }
}


src/Shape/Star.php
<?php

declare(strict_types=1);

namespace ImageGenerator\Shape;

use ImageGenerator\Canvas;
use ImageGenerator\Color;

final class Star extends Shape
{
    private int $points;

    private int $rotation;

    /**
     * Creates a new star.
     *
     * @param float $ratio The star size ratio. Must be between 0.0 and 1.0.
     * @param int $points The star number of points. Must be between 4 and 8.
     * @param int $rotation The star rotation angle in degrees. Must be between 0 and 360.
     */
    public function __construct(float $ratio, int $points, int $rotation)
    {
        $this->setRatio($ratio);
        $this->setPoints($points);
        $this->setRotation($rotation);
    }

    /**
     * Returns the star number of points.
     *
     * @return int
     */
    public function getPoints(): int
    {
        return $this->points;
    }

    /**
     * Sets the star number of points
     *
     * @param int $points The star number of points. Must be between 4 and 8.
     *
     * @return Star
     */
    public function setPoints(int $points): Star
    {
        if ($points < 4 || $points > 8) {
            throw new \InvalidArgumentException('The number of points must be between 4 and 8.');
        }

        $this->points = $points;
        return $this;
    }

    /**
     * Returns the star rotation angle.
     *
     * @return int
     */
    public function getRotation(): int
    {
        return $this->rotation;
    }

    /**
     * Sets the star rotation angle.
     *
     * @param int $rotation The star rotation angle in degrees. Must be between 0 and 360.
     *
     * @return Star
     */
    public function setRotation(int $rotation): Star
    {
        if ($rotation < 0 || $rotation > 360) {
            throw new \InvalidArgumentException('The rotation angle must be between 0 and 360.');
        }

        $this->rotation = $rotation;
        return $this;
    }

    /**
     * @internal
     */
    public function draw(Canvas $canvas, Color $color): void
    {
        $radius = (int)($this->getRatio() * $canvas->getGeometricMean() / 2);

        $x = rand(-$radius, $canvas->getWidth() + $radius);
        $y = rand(-$radius, $canvas->getHeight() + $radius);

        $sides = $this->getPoints() * 2;
        $points = [];

        for ($side = 0; $side < $sides; $side++) {
            $angle = deg2rad($this->getRotation() + 360 / $sides * $side);
            $distance = (int)($side % 2 === 0 ? $radius : $radius / 2);
            $points[] = $x + $distance * cos($angle);
            $points[] = $y + $distance * sin($angle);
        }

        $color = imagecolorallocatealpha(
            $canvas->getImage(),
            $color->getRed(),
            $color->getGreen(),
            $color->getBlue(),
            (int)(127 - $color->getAlpha() * 127)
        );

        if (false === $color) {
            throw new \RuntimeException('Could not allocate color.');
        }

        imagefilledpolygon($canvas->getImage(), $points, $color);
        imagecolordeallocate($canvas->getImage(), $color);
    }

    /**
     * Creates a random star.
     *
     * @param array{
     *     min_ratio?: float,
     *     max_ratio?: float,
     *     min_points?: int,
     *     max_points?: int,
     *     min_rotation?: int,
     *     max_rotation?: int
     * } $options An optional array of options.
     *
     * The following options are supported:
     *
     * min_ratio - The star minimum size ratio. Must be between 0.0 and 1.0. Defaults to 0.125.
     * max_ratio - The star maximum size ratio. Must be between 0.0 and 1.0. Defaults to 0.25.
     * min_points - The star minimum number of points. Must be between 4 and 8. Defaults to 4.
     * max_points - The star maximum number of points. Must be between 4 and 8. Defaults to 8.
     * min_rotation - The star minimum rotation angle in degrees. Must be between 0 and 360. Defaults to 0.
     * max_rotation - The star maximum rotation angle in degrees. Must be between 0 and 360. Defaults to 360.
     *
     * @return Star
     *
     * @phpstan-ignore-next-line
     */
    public static function random(array $options = []): Star
    {
        $minRatio = (int)(($options['min_ratio'] ?? 0.125) * 100);
        $maxRatio = (int)(($options['max_ratio'] ?? 0.25) * 100);
        $ratio = rand($minRatio, $maxRatio) / 100;

        $minPoints = $options['min_points'] ?? 4;
        $maxPoints = $options['max_points'] ?? 8;
        $points = rand($minPoints, $maxPoints);

        $minRotation = $options['min_rotation'] ?? 0;
        $maxRotation = $options['max_rotation'] ?? 360;
        $rotation = rand($minRotation, $maxRotation);

        return new Star($ratio, $points, $rotation);
    }
}


src/Shape/Shape.php
<?php

declare(strict_types=1);

namespace ImageGenerator\Shape;

use ImageGenerator\Canvas;
use ImageGenerator\Color;

abstract class Shape
{
    protected float $ratio;

    /**
     * Returns the shape size ratio.
     *
     * @return float
     */
    public function getRatio(): float
    {
        return $this->ratio;
    }

    /**
     * Sets the shape size ratio.
     *
     * @param float $ratio The shape size ratio. Must be between 0.0 and 1.0.
     *
     * @return self
     */
    public function setRatio(float $ratio): self
    {
        if ($ratio < 0 || $ratio > 1) {
            throw new \InvalidArgumentException('The size ratio must be between 0.0 and 1.0.');
        }

        $this->ratio = $ratio;
        return $this;
    }

    abstract public function draw(Canvas $canvas, Color $color): void;

    /**
     * Creates a random shape.
     *
     * @param array{
     *     circle?: array{
     *         min_ratio?: float,
     *         max_ratio?: float
     *     },
     *     polygon?: array{
     *         min_ratio?: float,
     *         max_ratio?: float,
     *         min_sides?: int,
     *         max_sides?: int,
     *         min_rotation?: int,
     *         max_rotation?: int
     *     },
     *     star?: array{
     *         min_ratio?: float,
     *         max_ratio?: float,
     *         min_points?: int,
     *         max_points?: int,
     *         min_rotation?: int,
     *         max_rotation?: int
     *     }
     * } $options An optional array of options for each shape.
     *
     * See each individual shape `random()` method for a list of supported options.
     *
     * @return Shape
     */
    public static function random(array $options = []): Shape
    {
        $shapes = ['circle', 'polygon', 'star'];
        $shape = $shapes[array_rand($shapes)];

        return match ($shape) {
            'circle' => Circle::random($options['circle'] ?? []),
            'polygon' => Polygon::random($options['polygon'] ?? []),
            'star' => Star::random($options['star'] ?? []),
        };
    }
}


src/Shape/Circle.php
<?php

declare(strict_types=1);

namespace ImageGenerator\Shape;

use ImageGenerator\Canvas;
use ImageGenerator\Color;

final class Circle extends Shape
{
    /**
     * Creates a new circle.
     *
     * @param float $ratio The circle size ratio. Must be between 0.0 and 1.0.
     */
    public function __construct(float $ratio)
    {
        $this->setRatio($ratio);
    }

    /**
     * @internal
     */
    public function draw(Canvas $canvas, Color $color): void
    {
        $radius = (int)($this->getRatio() * $canvas->getGeometricMean() / 2);

        $x = rand(-$radius, $canvas->getWidth() + $radius);
        $y = rand(-$radius, $canvas->getHeight() + $radius);

        $color = imagecolorallocatealpha(
            $canvas->getImage(),
            $color->getRed(),
            $color->getGreen(),
            $color->getBlue(),
            (int)(127 - $color->getAlpha() * 127)
        );

        if (false === $color) {
            throw new \RuntimeException('Could not allocate color.');
        }

        imagefilledellipse($canvas->getImage(), $x, $y, $radius * 2, $radius * 2, $color);
        imagecolordeallocate($canvas->getImage(), $color);
    }

    /**
     * Creates a random circle.
     *
     * @param array{min_ratio?: float, max_ratio?: float} $options An optional array of options.
     *
     * The following options are supported:
     *
     * min_ratio - The circle minimum size ratio. Must be between 0.0 and 1.0. Defaults to 0.125.
     * max_ratio - The circle maximum size ratio. Must be between 0.0 and 1.0. Defaults to 0.25.
     *
     * @return Circle
     *
     * @phpstan-ignore-next-line
     */
    public static function random(array $options = []): Circle
    {
        $minRatio = (int)(($options['min_ratio'] ?? 0.125) * 100);
        $maxRatio = (int)(($options['max_ratio'] ?? 0.25) * 100);
        $ratio = rand($minRatio, $maxRatio) / 100;

        return new Circle($ratio);
    }
}


src/Shape/Polygon.php
<?php

declare(strict_types=1);

namespace ImageGenerator\Shape;

use ImageGenerator\Canvas;
use ImageGenerator\Color;

final class Polygon extends Shape
{
    private int $sides;

    private int $rotation;

    /**
     * Creates a new polygon.
     *
     * @param float $ratio The polygon size ratio. Must be between 0.0 and 1.0.
     * @param int $sides The polygon number of sides. Must be between 3 and 12.
     * @param int $rotation The polygon rotation angle in degrees. Must be between 0 and 360.
     */
    public function __construct(float $ratio, int $sides, int $rotation)
    {
        $this->setRatio($ratio);
        $this->setSides($sides);
        $this->setRotation($rotation);
    }

    /**
     * Returns the polygon number of sides.
     *
     * @return int
     */
    public function getSides(): int
    {
        return $this->sides;
    }

    /**
     * Sets the polygon number of sides.
     *
     * @param int $sides The polygon number of sides. Must be between 3 and 12.
     *
     * @return Polygon
     */
    public function setSides(int $sides): Polygon
    {
        if ($sides < 3 || $sides > 12) {
            throw new \InvalidArgumentException('The number of sides must be between 3 and 12.');
        }

        $this->sides = $sides;
        return $this;
    }

    /**
     * Returns the polygon rotation angle.
     *
     * @return int
     */
    public function getRotation(): int
    {
        return $this->rotation;
    }

    /**
     * Sets the polygon rotation angle.
     *
     * @param int $rotation The polygon rotation angle in degrees. Must be between 0 and 360.
     *
     * @return Polygon
     */
    public function setRotation(int $rotation): Polygon
    {
        if ($rotation < 0 || $rotation > 360) {
            throw new \InvalidArgumentException('The rotation angle must be between 0 and 360.');
        }

        $this->rotation = $rotation;
        return $this;
    }

    /**
     * @internal
     */
    public function draw(Canvas $canvas, Color $color): void
    {
        $radius = (int)($this->getRatio() * $canvas->getGeometricMean() / 2);

        $x = rand(-$radius, $canvas->getWidth() + $radius);
        $y = rand(-$radius, $canvas->getHeight() + $radius);

        $points = [];

        for ($side = 0; $side < $this->getSides(); $side++) {
            $angle = deg2rad($this->getRotation() + 360 / $this->getSides() * $side);
            $points[] = $x + $radius * cos($angle);
            $points[] = $y + $radius * sin($angle);
        }

        $color = imagecolorallocatealpha(
            $canvas->getImage(),
            $color->getRed(),
            $color->getGreen(),
            $color->getBlue(),
            (int)(127 - $color->getAlpha() * 127)
        );

        if (false === $color) {
            throw new \RuntimeException('Could not allocate color.');
        }

        imagefilledpolygon($canvas->getImage(), $points, $color);
        imagecolordeallocate($canvas->getImage(), $color);
    }

    /**
     * Creates a random polygon.
     *
     * @param array{
     *     min_ratio?: float,
     *     max_ratio?: float,
     *     min_sides?: int,
     *     max_sides?: int,
     *     min_rotation?: int,
     *     max_rotation?: int
     * } $options An optional array of options.
     *
     * The following options are supported:
     *
     * min_radius - The polygon minimum size ratio. Must be between 0.0 and 1.0. Defaults to 0.125.
     * max_radius - The polygon maximum size ratio. Must be between 0.0 and 1.0. Defaults to 0.25.
     * min_sides - The polygon minimum number of sides. Must be between 3 and 12. Defaults to 3.
     * max_sides - The polygon maximum number of sides. Must be between 3 and 12. Defaults to 12.
     * min_rotation - The polygon minimum rotation angle in degrees. Must be between 0 and 360. Defaults to 0.
     * max_rotation - The polygon maximum rotation angle in degrees. Must be between 0 and 360. Defaults to 360.
     *
     * @return Polygon
     *
     * @phpstan-ignore-next-line
     */
    public static function random(array $options = []): Polygon
    {
        $minRatio = (int)(($options['min_ratio'] ?? 0.125) * 100);
        $maxRatio = (int)(($options['max_ratio'] ?? 0.25) * 100);
        $ratio = rand($minRatio, $maxRatio) / 100;

        $minSides = $options['min_sides'] ?? 3;
        $maxSides = $options['max_sides'] ?? 12;
        $sides = rand($minSides, $maxSides);

        $minRotation = $options['min_rotation'] ?? 0;
        $maxRotation = $options['max_rotation'] ?? 360;
        $rotation = rand($minRotation, $maxRotation);

        return new Polygon($ratio, $sides, $rotation);
    }
}


src/Canvas.php
<?php

namespace ImageGenerator;

use ImageGenerator\Shape\Shape;
use InvalidArgumentException;

final class Canvas
{
    private \GdImage $image;

    private int $width;

    private int $height;

    /**
     * Creates a new canvas.
     *
     * @param int $width The canvas width in pixels. Must be greater than 0.
     * @param int $height The canvas height in pixels. Must be greater than 0.
     */
    public function __construct(int $width, int $height)
    {
        if ($width <= 0) {
            throw new InvalidArgumentException('The width must be greater than 0.');
        }

        if ($height <= 0) {
            throw new InvalidArgumentException('The height must be greater than 0.');
        }

        $this->width = $width;
        $this->height = $height;

        $image = imagecreatetruecolor($width, $height);

        if (false === $image) {
            throw new \RuntimeException('Could not create the image');
        }

        $this->image = $image;
    }

    /**
     * Returns the canvas width.
     *
     * @return int
     */
    public function getWidth(): int
    {
        return $this->width;
    }

    /**
     * Returns the canvas height.
     *
     * @return int
     */
    public function getHeight(): int
    {
        return $this->height;
    }

    /**
     * Enables background transparency.
     *
     * @return Canvas
     */
    public function enableTransparency(): Canvas
    {
        $color = imagecolorallocatealpha($this->image, 0, 0, 0, 127);

        if (false === $color) {
            throw new \RuntimeException('Could not allocate color.');
        }

        imagefill($this->image, 0, 0, $color);
        imagesavealpha($this->image, true);

        return $this;
    }

    /**
     * Draws a shape on the canvas.
     *
     * @param Shape $shape The shape to draw.
     * @param Color $color The color used to draw the shape.
     *
     * @return Canvas
     */
    public function draw(Shape $shape, Color $color): Canvas
    {
        $shape->draw($this, $color);
        return $this;
    }

    /**
     * Fills the canvas with a given color.
     *
     * @param Color $color The color used to fill the canvas.
     *
     * @return Canvas
     */
    public function fill(Color $color): Canvas
    {
        $color = imagecolorallocatealpha(
            $this->image,
            $color->getRed(),
            $color->getGreen(),
            $color->getBlue(),
            (int)(127 - $color->getAlpha() * 127)
        );

        if (false === $color) {
            throw new \RuntimeException('Could not allocate color.');
        }

        imagefilledrectangle($this->image, 0, 0, $this->width, $this->height, $color);
        imagecolordeallocate($this->image, $color);

        return $this;
    }

    /**
     * Render the canvas to an image.
     *
     * @param string|null $file The destination file path. When omitted a temporary file is created.
     * @param string $format The image format. Must be one of 'avif', 'bmp', 'gif', 'jpeg', 'png', 'wbmp', 'webp' or 'xbm'. Defaults to 'png'.
     * @param mixed ...$options Additional options passed directly to the underlying image generation function.
     *
     * @return Image
     */
    public function render(string $file = null, string $format = 'png', mixed ...$options): Image
    {
        if (is_null($file)) {
            $file = tempnam(sys_get_temp_dir(), 'img');

            if ($file === false) {
                throw new \RuntimeException('Could not create temporary file.');
            }
        }

        switch ($format) {
            case 'avif':
                if (version_compare(PHP_VERSION, '8.1.0') < 0) {
                    throw new \RuntimeException('The AVIF format is available starting with PHP 8.1.');
                }

                $mimeType = image_type_to_mime_type(IMG_AVIF);
                imageavif($this->image, $file, ...$options);
                break;

            case 'bmp':
                $mimeType = image_type_to_mime_type(IMG_BMP);
                imagebmp($this->image, $file, ...$options);
                break;

            case 'gif':
                $mimeType = image_type_to_mime_type(IMG_GIF);
                imagegif($this->image, $file);
                break;

            case 'jpeg':
                $mimeType = image_type_to_mime_type(IMG_JPG);
                imagejpeg($this->image, $file, ...$options);
                break;

            case 'png':
                $mimeType = image_type_to_mime_type(IMG_PNG);
                imagepng($this->image, $file, ...$options);
                break;

            case 'wbmp':
                $mimeType = image_type_to_mime_type(IMG_WBMP);
                imagewbmp($this->image, $file);
                break;

            case 'webp':
                $mimeType = image_type_to_mime_type(IMG_WEBP);
                imagewebp($this->image, $file, ...$options);
                break;

            case 'xbm':
                $mimeType = image_type_to_mime_type(IMG_XPM);
                imagexbm($this->image, $file);
                break;

            default:
                throw new \LogicException(sprintf('Unsupported image format "%s".', $format));
        }

        return new Image($file, $this->width, $this->height, $mimeType);
    }

    /**
     * Creates a new canvas.
     *
     * @param int $width The canvas width in pixels. Must be greater than 0.
     * @param int $height The canvas height in pixels. Must be greater than 0.
     *
     * @return Canvas
     */
    public static function create(int $width, int $height): Canvas
    {
        return new Canvas($width, $height);
    }

    /**
     * @internal
     */
    public function getGeometricMean(): float
    {
        return sqrt($this->width * $this->height);
    }

    /**
     * @internal
     */
    public function getImage(): \GdImage
    {
        return $this->image;
    }
}
